File: app/src/main/java/org/tvheadend/tvhclient/ui/features/programs/ProgramListFragment.java
package org.tvheadend.tvhclient.ui.features.programs;

import android.app.SearchManager;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Filter;
import android.widget.ProgressBar;

import org.tvheadend.tvhclient.MainApplication;
import org.tvheadend.tvhclient.R;
import org.tvheadend.tvhclient.data.service.HtspService;
import org.tvheadend.tvhclient.domain.entity.Program;
import org.tvheadend.tvhclient.domain.entity.Recording;
import org.tvheadend.tvhclient.ui.base.BaseFragment;
import org.tvheadend.tvhclient.ui.base.callbacks.RecyclerViewClickCallback;
import org.tvheadend.tvhclient.ui.features.dialogs.GenreColorDialog;
import org.tvheadend.tvhclient.ui.features.dvr.RecordingAddEditActivity;
import org.tvheadend.tvhclient.ui.features.search.SearchRequestInterface;
import org.tvheadend.tvhclient.ui.features.search.StartSearchInterface;
import org.tvheadend.tvhclient.util.MiscUtils;
import org.tvheadend.tvhclient.util.menu.PopupMenuUtil;
import org.tvheadend.tvhclient.util.menu.SearchMenuUtils;

import java.util.List;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.widget.PopupMenu;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentTransaction;
import androidx.lifecycle.Lifecycle;
import androidx.lifecycle.ViewModelProviders;
import androidx.recyclerview.widget.DefaultItemAnimator;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;
import timber.log.Timber;

public class ProgramListFragment extends BaseFragment implements RecyclerViewClickCallback, LastProgramVisibleListener, SearchRequestInterface, Filter.FilterListener {

    private ProgramRecyclerViewAdapter recyclerViewAdapter;
    @BindView(R.id.recycler_view)
    protected RecyclerView recyclerView;
    @BindView(R.id.progress_bar)
    protected ProgressBar progressBar;

    private long selectedTime;
    private int selectedListPosition;
    private int channelId;
    private String channelName;
    private String searchQuery;
    private boolean loadingMoreProgramAllowed;
    private Runnable loadingProgramsAllowedTask;
    private final Handler loadingProgramAllowedHandler = new Handler();
    private Unbinder unbinder;
    private int programIdToBeEditedWhenBeingRecorded = 0;
    private boolean isSearchActive;
    private ProgramViewModel viewModel;

    public static ProgramListFragment newInstance(String channelName, int channelId, long selectedTime) {
        ProgramListFragment f = new ProgramListFragment();
        Bundle args = new Bundle();
        args.putString("channelName", channelName);
        args.putInt("channelId", channelId);
        args.putLong("selectedTime", selectedTime);
        f.setArguments(args);
        return f;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.recyclerview_fragment, container, false);
        unbinder = ButterKnife.bind(this, view);
        return view;
    }

    @Override
    public void onDestroyView() {
        recyclerView.setAdapter(null);
        super.onDestroyView();
        unbinder.unbind();
    }

    @Override
    public void onActivityCreated(Bundle savedInstanceState) {
        super.onActivityCreated(savedInstanceState);

        if (savedInstanceState != null) {
            channelId = savedInstanceState.getInt("channelId", 0);
            channelName = savedInstanceState.getString("channelName");
            selectedTime = savedInstanceState.getLong("selectedTime");
            selectedListPosition = savedInstanceState.getInt("listPosition", 0);
            searchQuery = savedInstanceState.getString(SearchManager.QUERY);
        } else {
            selectedListPosition = 0;

            Bundle bundle = getArguments();
            if (bundle != null) {
                channelId = bundle.getInt("channelId", 0);
                channelName = bundle.getString("channelName");
                selectedTime = bundle.getLong("selectedTime", System.currentTimeMillis());
                searchQuery = bundle.getString(SearchManager.QUERY);
            }
        }

        isSearchActive = !TextUtils.isEmpty(searchQuery);

        if (!isDualPane) {
            toolbarInterface.setTitle(isSearchActive ? getString(R.string.search_results) : channelName);
        }

        // Show the channel icons when a search is active and all channels shall be searched
        boolean showProgramChannelIcon = isSearchActive && channelId == 0;

        recyclerViewAdapter = new ProgramRecyclerViewAdapter(showProgramChannelIcon, this, this);
        recyclerView.setLayoutManager(new LinearLayoutManager(activity.getApplicationContext()));
        recyclerView.addItemDecoration(new DividerItemDecoration(activity.getApplicationContext(), LinearLayoutManager.VERTICAL));
        recyclerView.setItemAnimator(new DefaultItemAnimator());
        recyclerView.setAdapter(recyclerViewAdapter);

        recyclerView.setVisibility(View.GONE);
        progressBar.setVisibility(View.VISIBLE);

        viewModel = ViewModelProviders.of(activity).get(ProgramViewModel.class);
        if (!isSearchActive) {
            Timber.d("Search is not active, loading programs for channel " + channelName + " from time " + selectedTime);
            // A channel id and a channel name was given, load only the programs for the
            // specific channel and from the current time. Also load only those recordings
            // that belong to the given channel
            viewModel.getProgramsByChannelFromTime(channelId, selectedTime).observe(getViewLifecycleOwner(), this::handleObservedPrograms);
            viewModel.getRecordingsByChannelId(channelId).observe(getViewLifecycleOwner(), this::handleObservedRecordings);

            loadingMoreProgramAllowed = true;
            loadingProgramsAllowedTask = () -> loadingMoreProgramAllowed = true;

        } else {
            Timber.d("Search is active, loading programs from current time " + selectedTime);
            // No channel and channel name was given, load all programs
            // from the current time and all recordings from all channels
            viewModel.getProgramsFromTime(selectedTime).observe(getViewLifecycleOwner(), this::handleObservedPrograms);
            viewModel.getRecordings().observe(getViewLifecycleOwner(), this::handleObservedRecordings);

            loadingMoreProgramAllowed = false;
        }
    }

    private void handleObservedPrograms(List<Program> programs) {
        if (programs != null) {
            recyclerViewAdapter.addItems(programs);
        }
        if (isSearchActive) {
            if (activity instanceof StartSearchInterface) {
                ((StartSearchInterface) activity).startSearch();
            }
        }
        if (recyclerView != null) {
            recyclerView.setVisibility(View.VISIBLE);
        }
        if (progressBar != null) {
            progressBar.setVisibility(View.GONE);
        }

        if (!isDualPane) {
            if (!isSearchActive) {
                toolbarInterface.setSubtitle(activity.getResources().getQuantityString(R.plurals.items, recyclerViewAdapter.getItemCount(), recyclerViewAdapter.getItemCount()));
            } else {
                toolbarInterface.setSubtitle(activity.getResources().getQuantityString(R.plurals.programs, recyclerViewAdapter.getItemCount(), recyclerViewAdapter.getItemCount()));
            }
        }
        // Invalidate the menu so that the search menu item is shown in
        // case the adapter contains items now.
        activity.invalidateOptionsMenu();
    }

    /**
     * Check all recordings for the given channel to see if it belongs to a certain program
     * so the recording status of the particular program can be updated. This is required
     * because the programs are not updated automatically when recordings change.
     *
     * @param recordings The list of recordings
     */
    private void handleObservedRecordings(List<Recording> recordings) {
        if (recordings != null) {
            recyclerViewAdapter.addRecordings(recordings);
            for (Recording recording : recordings) {
                if (recording.getEventId() == programIdToBeEditedWhenBeingRecorded
                        && programIdToBeEditedWhenBeingRecorded > 0) {
                    programIdToBeEditedWhenBeingRecorded = 0;
                    Intent intent = new Intent(activity, RecordingAddEditActivity.class);
                    intent.putExtra("id", recording.getId());
                    intent.putExtra("type", "recording");
                    activity.startActivity(intent);
                    break;
                }
            }
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        loadingProgramAllowedHandler.removeCallbacks(loadingProgramsAllowedTask);
    }

    @Override
    public void onSaveInstanceState(@NonNull Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putInt("channelId", channelId);
        outState.putString("channelName", channelName);
        outState.putLong("selectedTime", selectedTime);
        outState.putInt("listPosition", selectedListPosition);
        outState.putString(SearchManager.QUERY, searchQuery);
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.program_list_options_menu, menu);
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        // Hide the genre color menu in dual pane mode or if no genre colors shall be shown
        final boolean showGenreColors = sharedPreferences.getBoolean("genre_colors_for_programs_enabled",
                activity.getResources().getBoolean(R.bool.pref_default_genre_colors_for_programs_enabled));

        MenuItem genreColorMenu = menu.findItem(R.id.menu_genre_color_info_programs);
        if (genreColorMenu != null) {
            genreColorMenu.setVisible(!isDualPane && showGenreColors);
        }

        if (!isSearchActive && isNetworkAvailable) {
            menu.findItem(R.id.menu_play).setVisible(true);
            menu.findItem(R.id.menu_cast).setVisible(MiscUtils.getCastSession(activity) != null);
        } else {
            menu.findItem(R.id.menu_play).setVisible(false);
            menu.findItem(R.id.menu_cast).setVisible(false);
        }
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case R.id.menu_play:
                menuUtils.handleMenuPlayChannel(channelId);
                return true;

            case R.id.menu_cast:
                return menuUtils.handleMenuCast("channelId", channelId);

            case R.id.menu_genre_color_info_programs:
                return GenreColorDialog.showDialog(activity);

            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void showProgramDetails(int position) {
        selectedListPosition = position;
        Program program = recyclerViewAdapter.getItem(position);
        if (program == null || !isVisible()
                || !activity.getLifecycle().getCurrentState().isAtLeast(Lifecycle.State.STARTED)) {
            return;
        }

        Fragment fragment = ProgramDetailsFragment.newInstance(program.getEventId(), program.getChannelId());
        FragmentTransaction ft = activity.getSupportFragmentManager().beginTransaction();
        ft.replace(R.id.main, fragment);
        ft.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE);
        ft.addToBackStack(null);
        ft.commit();
    }

    private void showPopupMenu(View view, int position) {
        final Program program = recyclerViewAdapter.getItem(position);
        if (activity == null || program == null) {
            return;
        }

        Recording recording = appRepository.getRecordingData().getItemByEventId(program.getEventId());

        PopupMenu popupMenu = new PopupMenu(activity, view);
        popupMenu.getMenuInflater().inflate(R.menu.program_popup_and_toolbar_menu, popupMenu.getMenu());
        popupMenu.getMenuInflater().inflate(R.menu.external_search_options_menu, popupMenu.getMenu());
        PopupMenuUtil.prepareMenu(activity, popupMenu.getMenu(), program, program.getRecording(), isNetworkAvailable, htspVersion, isUnlocked);
        PopupMenuUtil.prepareSearchMenu(popupMenu.getMenu(), program.getTitle(), isNetworkAvailable);

        popupMenu.setOnMenuItemClickListener(item -> {
            if (SearchMenuUtils.onMenuSelected(activity, item.getItemId(), program.getTitle(), program.getChannelId())) {
                return true;
            }
            switch (item.getItemId()) {
                case R.id.menu_record_stop:
                    return menuUtils.handleMenuStopRecordingSelection(recording, null);

                case R.id.menu_record_cancel:
                    return menuUtils.handleMenuCancelRecordingSelection(recording, null);

                case R.id.menu_record_remove:
                    return menuUtils.handleMenuRemoveRecordingSelection(recording, null);

                case R.id.menu_record_once:
                    return menuUtils.handleMenuRecordSelection(program.getEventId());

                case R.id.menu_record_once_and_edit:
                    programIdToBeEditedWhenBeingRecorded = program.getEventId();
                    return menuUtils.handleMenuRecordSelection(program.getEventId());

                case R.id.menu_record_once_custom_profile:
                    return menuUtils.handleMenuCustomRecordSelection(program.getEventId(), channelId);

                case R.id.menu_record_series:
                    return menuUtils.handleMenuSeriesRecordSelection(program.getTitle());

                case R.id.menu_play:
                    return menuUtils.handleMenuPlayChannel(channelId);

                case R.id.menu_cast:
                    return menuUtils.handleMenuCast("channelId", channelId);

                case R.id.menu_add_notification:
                    return menuUtils.handleMenuAddNotificationSelection(program);

                default:
                    return false;
            }
        });
        popupMenu.show();
    }

    public int getShownChannelId() {
        return channelId;
    }

    @Override
    public void onSearchRequested(String query) {
        recyclerViewAdapter.getFilter().filter(query, this);
    }

    @Override
    public boolean onSearchResultsCleared() {
        if (!TextUtils.isEmpty(searchQuery)) {
            searchQuery = "";
            recyclerViewAdapter.getFilter().filter("", this);
            return true;
        } else {
            return false;
        }
    }

    @Override
    public String getQueryHint() {
        return getString(R.string.search_programs);
    }

    @Override
    public void onLastProgramVisible(int position) {
        // Do not load more programs when a search query was given or all programs were loaded.
        if (isSearchActive || !loadingMoreProgramAllowed || !isNetworkAvailable) {
            return;
        }

        loadingMoreProgramAllowed = false;
        loadingProgramAllowedHandler.postDelayed(loadingProgramsAllowedTask, 2000);

        Program lastProgram = recyclerViewAdapter.getItem(position);
        Timber.d("Loading more programs after " + lastProgram.getTitle());

        Intent intent = new Intent(activity, HtspService.class);
        intent.setAction("getEvents");
        intent.putExtra("eventId", lastProgram.getNextEventId());
        intent.putExtra("channelId", lastProgram.getChannelId());
        intent.putExtra("channelName", channelName);
        intent.putExtra("numFollowing", 25);
        intent.putExtra("showMessage", true);

        if (MainApplication.isActivityVisible()) {
            activity.startService(intent);
        }
    }

    @Override
    public void onFilterComplete(int count) {
        if (!isDualPane) {
            if (!isSearchActive) {
                toolbarInterface.setSubtitle(activity.getResources().getQuantityString(R.plurals.items, recyclerViewAdapter.getItemCount(), recyclerViewAdapter.getItemCount()));
            } else {
                toolbarInterface.setSubtitle(activity.getResources().getQuantityString(R.plurals.programs, recyclerViewAdapter.getItemCount(), recyclerViewAdapter.getItemCount()));
            }
        }
    }

    @Override
    public void onClick(View view, int position) {
        showProgramDetails(position);
    }

    @Override
    public boolean onLongClick(View view, int position) {
        showPopupMenu(view, position);
        return true;
    }

    public void updatePrograms(long selectedTime) {
        this.selectedTime = selectedTime;
        viewModel.getProgramsByChannelFromTime(channelId, selectedTime).observe(getViewLifecycleOwner(), this::handleObservedPrograms);
    }
}

File: app/src/main/java/org/tvheadend/tvhclient/ui/features/programs/ProgramListFragment.kt
package org.tvheadend.tvhclient.ui.features.programs

import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.Filter
import androidx.appcompat.widget.PopupMenu
import androidx.fragment.app.FragmentTransaction
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.LinearLayoutManager
import kotlinx.android.synthetic.main.recyclerview_fragment.*
import org.tvheadend.data.entity.Recording
import org.tvheadend.tvhclient.R
import org.tvheadend.tvhclient.service.HtspService
import org.tvheadend.tvhclient.ui.base.BaseFragment
import org.tvheadend.tvhclient.ui.common.*
import org.tvheadend.tvhclient.ui.common.interfaces.ClearSearchResultsOrPopBackStackInterface
import org.tvheadend.tvhclient.ui.common.interfaces.RecyclerViewClickInterface
import org.tvheadend.tvhclient.ui.common.interfaces.SearchRequestInterface
import org.tvheadend.tvhclient.util.extensions.getCastSession
import org.tvheadend.tvhclient.util.extensions.gone
import org.tvheadend.tvhclient.util.extensions.visible
import org.tvheadend.tvhclient.util.extensions.visibleOrGone
import timber.log.Timber

class ProgramListFragment : BaseFragment(), RecyclerViewClickInterface, LastProgramVisibleListener, SearchRequestInterface, Filter.FilterListener, ClearSearchResultsOrPopBackStackInterface {

    lateinit var recyclerViewAdapter: ProgramRecyclerViewAdapter
    private lateinit var programViewModel: ProgramViewModel
    private var loadingMoreProgramAllowed: Boolean = false
    private var programIdToBeEditedWhenBeingRecorded = 0
    private var lastProgramItemCount = 0
    private var channelId = 0

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(R.layout.recyclerview_fragment, container, false)
    }

    override fun onActivityCreated(savedInstanceState: Bundle?) {
        super.onActivityCreated(savedInstanceState)
        programViewModel = ViewModelProvider(requireActivity()).get(ProgramViewModel::class.java)

        arguments?.let {
            programViewModel.channelId = it.getInt("channelId", 0)
            programViewModel.channelIdLiveData.value = it.getInt("channelId", 0)
            programViewModel.selectedTimeLiveData.value = it.getLong("selectedTime", System.currentTimeMillis())
            programViewModel.channelName = it.getString("channelName", "")
        }

        // Show the channel icons when a search is active and all channels shall be searched
        programViewModel.showProgramChannelIcon = baseViewModel.isSearchActive && channelId == 0

        recyclerViewAdapter = ProgramRecyclerViewAdapter(programViewModel, this, this)
        recycler_view.layoutManager = LinearLayoutManager(activity)
        recycler_view.adapter = recyclerViewAdapter
        recycler_view.gone()
        search_progress?.visibleOrGone(baseViewModel.isSearchActive)

        Timber.d("Observing programs")
        programViewModel.programs.observe(viewLifecycleOwner,  { progs ->
            if (progs != null) {
                Timber.d("View model returned ${progs.size} programs")
                recyclerViewAdapter.addItems(progs.toMutableList())
                observeSearchQuery()
                observeRecordings()
            }

            recycler_view?.visible()
            showStatusInToolbar()
            activity?.invalidateOptionsMenu()
        })

        Timber.d("Observing channel id")
        programViewModel.channelIdLiveData.observe(viewLifecycleOwner,  { id ->
            if (id != null) {
                Timber.d("View model returned channel id $id")
                channelId = id
            }
        })
    }

    private fun observeRecordings() {
        Timber.d("Observing recordings")
        programViewModel.recordings.observe(viewLifecycleOwner,  { recs ->
            if (recs != null) {
                Timber.d("View model returned ${recs.size} recordings")
                handleObservedRecordings(recs)
            }
        })
    }

    private fun observeSearchQuery() {
        Timber.d("Observing search query")
        baseViewModel.searchQueryLiveData.observe(viewLifecycleOwner,  { query ->
            loadingMoreProgramAllowed = if (query.isNotEmpty()) {
                Timber.d("View model returned search query '$query'")
                onSearchRequested(query)
                false

            } else {
                Timber.d("View model returned empty search query")
                onSearchResultsCleared()
                true
            }
        })
    }

    /**
     * Check all recordings for the given channel to see if it belongs to a certain program
     * so the recording status of the particular program can be updated. This is required
     * because the programs are not updated automatically when recordings change.
     *
     * @param recordings The list of recordings
     */
    private fun handleObservedRecordings(recordings: List<Recording>) {
        recyclerViewAdapter.addRecordings(recordings)
        for (recording in recordings) {
            if (recording.eventId == programIdToBeEditedWhenBeingRecorded && programIdToBeEditedWhenBeingRecorded > 0) {
                programIdToBeEditedWhenBeingRecorded = 0
                editSelectedRecording(requireActivity(), recording.id)
                break
            }
        }
    }

    override fun onCreateOptionsMenu(menu: Menu, inflater: MenuInflater) {
        super.onCreateOptionsMenu(menu, inflater)
        inflater.inflate(R.menu.program_list_options_menu, menu)
    }

    override fun onPrepareOptionsMenu(menu: Menu) {
        val ctx = context ?: return
        val showGenreColors = sharedPreferences.getBoolean("genre_colors_for_programs_enabled", resources.getBoolean(R.bool.pref_default_genre_colors_for_programs_enabled))
        // Hide the genre color menu in dual pane mode or if no genre colors shall be shown
        menu.findItem(R.id.menu_genre_color_information)?.isVisible = !isDualPane && showGenreColors

        if (!baseViewModel.isSearchActive && isConnectionToServerAvailable) {
            menu.findItem(R.id.menu_play)?.isVisible = true
            menu.findItem(R.id.menu_cast)?.isVisible = ctx.getCastSession() != null
        } else {
            menu.findItem(R.id.menu_play)?.isVisible = false
            menu.findItem(R.id.menu_cast)?.isVisible = false
        }
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val ctx = context ?: return super.onOptionsItemSelected(item)
        return when (item.itemId) {
            R.id.menu_play -> playSelectedChannel(ctx, channelId, isUnlocked)
            R.id.menu_cast -> castSelectedChannel(ctx, channelId)
            R.id.menu_genre_color_information -> showGenreColorDialog(ctx)
            else -> super.onOptionsItemSelected(item)
        }
    }

    private fun showProgramDetails(position: Int) {
        val program = recyclerViewAdapter.getItem(position)
        if (program == null
                || !isVisible
                || !lifecycle.currentState.isAtLeast(Lifecycle.State.STARTED)) {
            return
        }

        val fragment = ProgramDetailsFragment.newInstance(program.eventId, program.channelId)
        activity?.supportFragmentManager?.beginTransaction()?.also {
            it.replace(R.id.main, fragment)
            it.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_FADE)
            it.addToBackStack(null)
            it.commit()
        }
    }

    private fun showPopupMenu(view: View, position: Int) {
        val ctx = context ?: return
        val program = recyclerViewAdapter.getItem(position) ?: return

        val popupMenu = PopupMenu(ctx, view)
        popupMenu.menuInflater.inflate(R.menu.program_popup_and_toolbar_menu, popupMenu.menu)
        popupMenu.menuInflater.inflate(R.menu.external_search_options_menu, popupMenu.menu)

        preparePopupOrToolbarRecordingMenu(ctx, popupMenu.menu, program.recording, isConnectionToServerAvailable, htspVersion, isUnlocked)
        preparePopupOrToolbarSearchMenu(popupMenu.menu, program.title, isConnectionToServerAvailable)
        preparePopupOrToolbarMiscMenu(ctx, popupMenu.menu, program, isConnectionToServerAvailable, isUnlocked)

        popupMenu.setOnMenuItemClickListener { item ->
            when (item.itemId) {
                R.id.menu_stop_recording -> return@setOnMenuItemClickListener showConfirmationToStopSelectedRecording(ctx, program.recording, null)
                R.id.menu_cancel_recording -> return@setOnMenuItemClickListener showConfirmationToCancelSelectedRecording(ctx, program.recording, null)
                R.id.menu_remove_recording -> return@setOnMenuItemClickListener showConfirmationToRemoveSelectedRecording(ctx, program.recording, null)
                R.id.menu_record_program -> return@setOnMenuItemClickListener recordSelectedProgram(ctx, program.eventId, programViewModel.getRecordingProfile(), htspVersion)
                R.id.menu_record_program_and_edit -> {
                    programIdToBeEditedWhenBeingRecorded = program.eventId
                    return@setOnMenuItemClickListener recordSelectedProgram(ctx, program.eventId, programViewModel.getRecordingProfile(), htspVersion)
                }
                R.id.menu_record_program_with_custom_profile -> return@setOnMenuItemClickListener recordSelectedProgramWithCustomProfile(ctx, program.eventId, program.channelId, programViewModel.getRecordingProfileNames(), programViewModel.getRecordingProfile())
                R.id.menu_record_program_as_series_recording -> return@setOnMenuItemClickListener recordSelectedProgramAsSeriesRecording(ctx, program.title, programViewModel.getRecordingProfile(), htspVersion)
                R.id.menu_play -> return@setOnMenuItemClickListener playSelectedChannel(ctx, channelId, isUnlocked)
                R.id.menu_cast -> return@setOnMenuItemClickListener castSelectedChannel(ctx, channelId)

                R.id.menu_search_imdb -> return@setOnMenuItemClickListener searchTitleOnImdbWebsite(ctx, program.title)
                R.id.menu_search_fileaffinity -> return@setOnMenuItemClickListener searchTitleOnFileAffinityWebsite(ctx, program.title)
                R.id.menu_search_youtube -> return@setOnMenuItemClickListener searchTitleOnYoutube(ctx, program.title)
                R.id.menu_search_google -> return@setOnMenuItemClickListener searchTitleOnGoogle(ctx, program.title)
                R.id.menu_search_epg -> return@setOnMenuItemClickListener searchTitleInTheLocalDatabase(requireActivity(), baseViewModel, program.title, program.channelId)

                R.id.menu_add_notification -> return@setOnMenuItemClickListener addNotificationProgramIsAboutToStart(ctx, program, programViewModel.getRecordingProfile())
                else -> return@setOnMenuItemClickListener false
            }
        }
        popupMenu.show()
    }

    override fun onSearchRequested(query: String) {
        recyclerViewAdapter.filter.filter(query, this)
    }

    override fun onSearchResultsCleared() {
        recyclerViewAdapter.filter.filter("", this)
    }

    override fun getQueryHint(): String {
        return getString(R.string.search_programs)
    }

    override fun onLastProgramVisible(position: Int) {
        // Do not load more programs when a search query was given or all programs were loaded.
        if (baseViewModel.isSearchActive
                || !loadingMoreProgramAllowed
                || !isConnectionToServerAvailable
                || recyclerViewAdapter.itemCount == lastProgramItemCount) {
            return
        }
        lastProgramItemCount = recyclerViewAdapter.itemCount

        val lastProgram = recyclerViewAdapter.getItem(position)
        lastProgram?.let {
            Timber.d("Loading more programs after ${lastProgram.title}")

            val intent = Intent(activity, HtspService::class.java)
            intent.action = "getEvents"
            intent.putExtra("eventId", lastProgram.nextEventId)
            intent.putExtra("channelId", lastProgram.channelId)
            intent.putExtra("channelName", programViewModel.channelName)
            intent.putExtra("numFollowing", 25)
            intent.putExtra("showMessage", true)

            activity?.startService(intent)
        }
    }

    override fun onFilterComplete(count: Int) {
        showStatusInToolbar()
        search_progress?.gone()
    }

    private fun showStatusInToolbar() {
        context?.let {
            if (!isDualPane) {
                if (!baseViewModel.isSearchActive) {
                    toolbarInterface.setTitle(programViewModel.channelName)
                    toolbarInterface.setSubtitle(it.resources.getQuantityString(R.plurals.programs, recyclerViewAdapter.itemCount, recyclerViewAdapter.itemCount))
                } else {
                    toolbarInterface.setTitle(getString(R.string.search_results))
                    toolbarInterface.setSubtitle(it.resources.getQuantityString(R.plurals.programs, recyclerViewAdapter.itemCount, recyclerViewAdapter.itemCount))
                }
            } else {
                val toolbarTitle = "${programViewModel.channelName}: ${it.resources.getQuantityString(R.plurals.items, recyclerViewAdapter.itemCount, recyclerViewAdapter.itemCount)}"
                toolbarInterface.setSubtitle(toolbarTitle)
            }
        }
    }

    override fun onClick(view: View, position: Int) {
        showProgramDetails(position)
    }

    override fun onLongClick(view: View, position: Int): Boolean {
        showPopupMenu(view, position)
        return true
    }

    companion object {

        fun newInstance(channelName: String = "", channelId: Int = 0, selectedTime: Long = System.currentTimeMillis()): ProgramListFragment {
            val f = ProgramListFragment()
            val args = Bundle()
            args.putString("channelName", channelName)
            args.putInt("channelId", channelId)
            args.putLong("selectedTime", selectedTime)
            f.arguments = args
            return f
        }
    }
}
